x-pack/plugin/profiling/src/main/java/org/elasticsearch/xpack/profiling/TransportGetStackTracesAction.java

```diff
@@ -208,42 +232,78 @@ private void searchEventGroupedByStackTrace(
+                    .size(maxSupportedUniqueHosts)
+                    .field("host.id")
                     // 'execution_hint: map' skips the slow building of ordinals that we don't need.
                     // Especially with high cardinality fields, this makes aggregations really slow.
                     .executionHint("map")
-                    .subAggregation(new SumAggregationBuilder("count").field("Stacktrace.count"))
+                    .subAggregation(
+                        new TermsAggregationBuilder("group_by")
+                            // 'size' should be max 100k, but might be slightly more. Better be on the safe side.
+                            .size(150_000)
+                            .field("Stacktrace.id")
+                            // 'execution_hint: map' skips the slow building of ordinals that we don't need.
+                            // Especially with high cardinality fields, this makes aggregations really slow.
+                            .executionHint("map")
+                            .subAggregation(new SumAggregationBuilder("count").field("Stacktrace.count"))
+                    )
             )
             .addAggregation(new SumAggregationBuilder("total_count").field("Stacktrace.count"))
             .execute(handleEventsGroupedByStackTrace(client, responseBuilder, submitListener, searchResponse -> {
-                StringTerms stacktraces = searchResponse.getAggregations().get("group_by");
-                Sum totalCountAgg = searchResponse.getAggregations().get("total_count");
-                long totalCount = Math.round(totalCountAgg.value());
+                long totalCount = getLongFromResponse(searchResponse, "total_count");
+
                 Resampler resampler = new Resampler(request, responseBuilder.getSamplingRate(), totalCount);
-                // sort items lexicographically to access Lucene's term dictionary more efficiently when issuing an mget request.
+                StringTerms hosts = searchResponse.getAggregations().get("group_by");
+
+                // Sort items lexicographically to access Lucene's term dictionary more efficiently when issuing an mget request.
                 // The term dictionary is lexicographically sorted and using the same order reduces the number of page faults
                 // needed to load it.
                 long totalFinalCount = 0;
-                Map<String, Long> stackTraceEvents = new TreeMap<>();
-                for (StringTerms.Bucket bucket : stacktraces.getBuckets()) {
-                    Sum count = bucket.getAggregations().get("count");
-                    int finalCount = resampler.adjustSampleCount((int) count.value());
-                    totalFinalCount += finalCount;
-                    if (finalCount > 0) {
-                        stackTraceEvents.put(bucket.getKeyAsString(), (long) finalCount);
+                List<HostEventCount> hostEventCounts = new ArrayList<>(150_000);
```

Comment:
I suppose that value matches the max bucket count in the aggregation. Can we extract this as a constant?